src/engine/TPA.cc

```diff
@@ -1403,74 +1425,77 @@ VerificationResult TransitionSystemNetworkManager::solve() && {
             if (current == getChainStart()) {
                 return {VerificationAnswer::SAFE, ValidityWitness{}};
             }
             assert(getNode(current).trulySafe != PTRef_Undef);
-//            std::cout << "Query not reachable in " << current.id << ". These states are safe: " << logic.pp(explanation) << std::endl;
             getNode(current).trulySafe = logic.mkOr(getNode(current).trulySafe, explanation);
-            EId previousEdge = getIncomingEdge(current);
-            auto & previousNode = getNode(graph.getSource(previousEdge));
+            auto previousEdge = activePath.last();
+            auto &previousNode = getNode(graph.getSource(previousEdge));
             assert(previousNode.trulyReached != PTRef_Undef);
             PTRef updatedConditions = logic.mkNot(getNode(current).trulySafe);
             auto [edgeRes, edgeExplanation] = queryEdge(previousEdge, previousNode.trulyReached, updatedConditions);
             if (reachable(edgeRes)) { // New reached, not refuted yet, states
-//                std::cout << "Found new potential initial states: " << logic.pp(edgeExplanation) << std::endl;
                 getNode(current).solver->resetInitialStates(edgeExplanation);
                 continue; // Repeat the query for the same TS with new initial states
             } else { // Cannot continue from currently computed truly reached states
                 previousNode.trulyReached = PTRef_Undef;
-//                std::cout << "Strenghtening source TS with: " << logic.pp(edgeExplanation) << std::endl;
-                previousNode.solver->updateQueryStates(logic.mkNot(edgeExplanation));
+                addRestrictions(graph.getSource(previousEdge), logic.mkNot(edgeExplanation));
+                activePath.pop();
                 current = graph.getSource(previousEdge);
                 continue; // Blocked states have been propagated to the previous node, repeat the query there.
             }
         }
-        assert(false); // Unreachable
     }
     throw std::logic_error("Unreachable!");
 }
 
 InvalidityWitness TransitionSystemNetworkManager::computeInvalidityWitness() const {
     std::vector<EId> errorPath;
-    auto current = getChainStart();
-    errorPath.push_back(getIncomingEdge(current));
-    while (true) {
+    SymRef current;
+    for(int i = 0; i < activePath.size(); i++){
+        EId next = activePath[i];
+        errorPath.push_back(next);
+        current = graph.getTarget(next);
         auto steps = getNode(current).solver->getTransitionStepCount();
         errorPath.insert(errorPath.end(), steps, getSelfLoopFor(current, graph, adjacencyRepresentation).value());
         if (current == getChainEnd()) {
             break;
         }
-        EId next = getOutgoingEdge(current);
-        errorPath.push_back(next);
-        current = graph.getTarget(next);
+
     }
-    errorPath.push_back(getNode(current).next);
+    errorPath.push_back(getNode(current).children[0]);
```

Comment:
Error path should be built based on the activePath, not just taking the first child.

Reply:
It is built based on the active path, that's just the connection to the last element of the chain. At this stage `current == getChainEnd()`